ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshState.java

```diff
@@ -0,0 +1,222 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.p2p.discovery.internal;
+
+import static java.util.stream.Collectors.toList;
+import static tech.pegasys.pantheon.ethereum.p2p.discovery.internal.PeerDistanceCalculator.distance;
+
+import tech.pegasys.pantheon.ethereum.p2p.peers.Peer;
+import tech.pegasys.pantheon.ethereum.p2p.peers.PeerBlacklist;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import io.vertx.core.Vertx;
+
+class RecursivePeerRefreshState {
+  private final int CONCURRENT_REQUEST_LIMIT = 3;
+  private final int TIMEOUT_TASK_DELAY = 30000; // 30 Seconds
+  private final BytesValue target;
+  private final PeerBlacklist peerBlacklist;
+  private final BondingAgent bondingAgent;
+  private final NeighborFinder neighborFinder;
+  private final List<PeerDistance> anteList;
+  private final List<OutstandingRequest> outstandingRequestList;
+  private final List<BytesValue> contactedInCurrentExecution;
+  private final Vertx vertx;
+
+  RecursivePeerRefreshState(
+      final BytesValue target,
+      final PeerBlacklist peerBlacklist,
+      final BondingAgent bondingAgent,
+      final NeighborFinder neighborFinder,
+      final Vertx vertx) {
+    this.target = target;
+    this.peerBlacklist = peerBlacklist;
+    this.bondingAgent = bondingAgent;
+    this.neighborFinder = neighborFinder;
+    this.vertx = vertx;
+    this.anteList = new ArrayList<>();
+    this.outstandingRequestList = new ArrayList<>();
+    this.contactedInCurrentExecution = new ArrayList<>();
+
+    commenceTimeoutTask();
+  }
+
+  private void commenceTimeoutTask() {
+    vertx.setPeriodic(
+        TIMEOUT_TASK_DELAY,
+        v -> {
+          List<OutstandingRequest> outstandingRequestListCopy =
+              new ArrayList<>(outstandingRequestList);
+
+          for (OutstandingRequest outstandingRequest : outstandingRequestListCopy) {
+            if (outstandingRequest.isExpired()) {
+              List<Peer> queryCandidates = determineFindNodeCandidates(anteList.size());
+              for (Peer candidate : queryCandidates) {
+                if (!contactedInCurrentExecution.contains(candidate.getId())
+                    && !outstandingRequestList.contains(new OutstandingRequest(candidate))) {
+                  executeFindNodeRequest(candidate);
+                }
+              }
+              outstandingRequestList.remove(outstandingRequest);
+            }
+          }
+        });
```

Comment:
nit: Most of the variables in here should be final.  I suggest enabling IntelliJ's inspections "Local variable or parameter can be final" and "Field may be final".